### DISCUSSION TITLE
How to activate Ulauncher when the hotkey includes a letter and Caps Lock is on?

### DISCUSSION THREAD

**Discussion by leesiongchan** [Ulauncher/Ulauncher#910] [Q&A] [answered]
### Steps to reproduce
1. Set HotKey to Alt+S
2. Turn on Caps Lock
3. Try activate ulauncher with Alt+S

### Current behavior
ulauncher not activated

### Expected behavior
ulauncher should be opened

### Environment
**Ulauncher version**: 5.7.5
**Linux distro**: Arch Linux 
**Desktop environment**: Gnome 3

**Comment by troycurtisjr** [comment, marked as answer]
The library we use implements it like this.

As a workaround you could set a custom Gnome keyboard shortcut for `Alt-s` with the command `ulauncher-toggle` to allow it to work as you expect.

If you don't get focus on the new window, when you use `ulauncher-toggle`, make sure `wmctrl` is installed.

**Reply by RedBearAK** [reply]
@friday @troycurtisjr 

> Isn't that actually 2 hotkeys? One is `Alt-s` and another is `Alt-S` (when Caps Lock is on). It doesn't seem like you'd generally want to automatically be case insensitive for the hotkeys as that could lead to unexpected behavior in other usages.

This is incorrect. `Alt+s` and `Alt+Shift+S` are different keyboard shortcuts, if you want to represent them that way. But `Alt+s` and `Alt+S` (with Caps Lock enabled) are literally just the **same two physical keys on the keyboard**. That is what is important to the user being able to memorize and rely on the response to a specific keyboard shortcut under all circumstances. 

I've never in over three decades been on any type of computer system that responded to shortcuts differently when the Caps Lock key is on, as if you were holding down the Shift key. That is not a logical way to detect and respond to keyboard shortcut combinations. It would completely upend the reaction to all keyboard shortcuts involving letters, which is most of them, whenever the Caps Lock state happened to be on. I've never experienced anything like this on any operating system. No one would use keyboard shortcuts if they were that unreliable just because Caps Lock was on. The Caps Lock state should only be relevant for typing things that are not keyboard shortcuts (i.e., not involving modifier keys). 

Whatever library is detecting the keyboard shortcuts should be explicitly testing for whether the user is holding down the Shift key, and reacting to uppercase letters in exactly the same way as lowercase letters if the Shift key isn't being held as part of the shortcut combo. Or remapping uppercase letters to lowercase before even evaluating the shortcut. 

Responding to `Alt+s` and `Alt+S` differently, as if one included the Shift key even though the Shift key code was not detected, is not rational. 

When you set up a shortcut in GNOME's keyboard settings with the Caps Lock on, it does unexpectedly create the shortcut as if you were holding down the Shift key, because it sees a capital letter. But this is obviously a bug in the shortcut creation detection logic, since it should also be detecting that you weren't actually holding down the Shift key. And if you try to use that shortcut with Caps Lock enabled but no real Shift key held? Guess what. It doesn't work. Which is correct. The shell treats the uppercase letter issued by the enabled Caps Lock state as if it was lowercase, or at least properly acknowledges that it doesn't detect the Shift key code, so the shortcut doesn't activate. Add the real Shift key and it works, whether Caps Lock is on or off. 

I made a shortcut to open a new Nautilus window in a specific location (e.g., `nautilus --new-window /home/myuser/Videos`) to test this. With Caps Lock enabled, the GNOME keyboard shortcut dialog showed a shortcut of Shift+Ctrl+Alt+Super+V after I pressed only Ctrl+Alt+Super+V to set the shortcut. But if I try to actually use the shortcut by just pressing Ctrl+Alt+Super+V without adding Shift, it doesn't activate the action, regardless of Caps Lock status. The Caps Lock state is ignored/irrelevant. 

So that "solution" you offered would result in a keyboard shortcut of `Alt+Shift+S` if the user tried to create an `Alt+S` shortcut with the Caps Lock on. It wouldn't work, because GNOME doesn't treat uppercase letters differently without the Shift key.

**Reply by friday** [reply]
> Responding to `Alt+s` and `Alt+S` differently, as if one included the Shift key even though the Shift key code was not detected, is not rational.

I actually agree, but Troys suggestion is still the correct solution because [the library we use](https://github.com/kupferlauncher/keybinder) implements it to treat caps like this, and there's to my knowing no other library we can use, especially with Wayland support. So this workaround is the best we can do for now.

**Reply by RedBearAK** [reply]
I've always used something like Alt+Space (well, Ctrl+Space remapped to Alt+F1 with Kinto.sh, but whatever, it works and it's not affected by Caps Lock state), so I've never run into this, but it is definitely an issue, and a behavior that I've never seen in any other piece of software. This is in Ubuntu 21.10. 

Strangely, if you have Caps Lock on while setting the shortcut to `Alt+S`, it will work when Caps Lock is on and then refuse to work when Caps Lock is off. Set it with Caps Lock off and it will work only when Caps Lock is off. But in either case, the settings dialog will display `Alt+S` with an uppercase "S". Key combos involving Shift do not seem to work at all, no matter what I try, even using long combos to make sure I'm not clashing with some existing system shortcut. And that's regardless of Caps Lock state. 

This library seems to have pretty serious problems. I sure hope you're going with something else for v6.0. I guess I'm lucky that I was trained by Spotlight to use the equivalent of Command+Space.

**Reply by friday** [reply]
I removed the opinion part and only included the workaround part (sorry @troycurtisjr). This is my bad for converting from an issue to an "Answer", but the issue couldn't be solved, and it had a good workaround, so it was better for this format.

I also verified that the suggestion works, and it does. You just have to make sure not to have caps on when you add the command.

**Reply by RedBearAK** [reply]
> I also verified that the suggestion works, and it does. You just have to make sure not to have caps on when you add the command.

I suppose that could be considered the technically correct corollary to what I said about it NOT working if you try to create the shortcut with Caps Lock on, but I have tried this and keep running into the issue that the Ulauncher dialog does not have the keyboard focus when it appears, if it is called via this new system shortcut. It may just be a quirk of my particular system, but I've tried it repeatedly and so far the only way I've been able to reliably get Ulauncher to appear with a blinking cursor and accept keyboard input (without clicking on the dialog with a mouse first) is by hiding all windows until only the desktop is visible. This behavior is in contrast from the behavior when activating Ulauncher with the shortcut set in the Ulauncher preferences, which pretty much always results in the Ulauncher dialog having keyboard focus. (But not 100% of the time, just almost always.) 

Maybe on other systems this technique (activation of `ulauncher-toggle` with a system shortcut) works better, but on my machine I wouldn't really consider this a "solution" if I had to use it. Just FYI. The window that has keyboard focus when I activate the shortcut always seems to retain the focus. 

Again, Ubuntu 21.10, so GNOME shell v40.5. 

I've tried adding `&` to the command and using absolute path versus just the command name. No difference.

**Reply by troycurtisjr** [reply]
For the focus issue, you might confirm that you have wmctrl installed. The toggle script uses that in order to give ulauncher the focus.

**Reply by RedBearAK** [reply]
Confirmed that it was NOT installed. The dialog reliably has keyboard focus after I installed `wmctrl`. Kinda feel like that should be marked as a necessary dependency, or at the very least a "recommended" dependency when installing Ulauncher. The launcher dialog is fairly useless as an effective launcher without it. 

Thanks.

**Comment by friday** [comment]
Libkeybinder which is handling this for us is something we eventually will have to replace for Wayland support, and in the meanwhile issues like this will remain.
